### ISSUE TITLE
Add stepwise scheduling for GradientAccumulationScheduler

### ISSUE THREAD

**Issue by lgienapp** [PyTorchLightning/pytorch-lightning#21534]
### Description & Motivation

Currently, the GradientAccumulationScheduler only supports scheduling on epoch intervals. However, during pretraining tasks, the model might only run for a single epoch. Therefore, it would be beneficial to be able to schedule the gradient accumulation according to `trainer.global_step` taken. 

Proposal:
- add a `interval` parameter to GradientAccumulationScheduler, which can be `"epoch"` or `"step"`, defaulting to `"epoch"` for backwards compatibility
- add a condition to the current `on_train_epoch_start` to only trigger if `interval == "epoch"` 
- add an `on_train_batch_start`/`on_after_optimizer_step` hook, triggering if `interval == "step"` 

However, given the current warning of scheduling being incompatible with DeepSpeed, I am not sure if scheduling on steps would be unsupported by all/some strategies.

### Pitch

I want to be able to scheduling gradient accumulation by `trainer.global_step` instead of `trainer.current_epoch`.

### Additional context

Could depend on having an `on_optimizer_step` hook for callbacks. See https://github.com/Lightning-AI/pytorch-lightning/issues/11688#issuecomment-1812863621

cc @lantiga

**Comment by xXMrNidaXx** [comment]
Stepwise scheduling for GradientAccumulationScheduler would be super useful!

**Proposed API:**
```python
from lightning.pytorch.callbacks import GradientAccumulationScheduler

# Current (epoch-based)
accumulator = GradientAccumulationScheduler(scheduling={0: 8, 4: 4, 8: 1})

# Proposed (step-based)
accumulator = GradientAccumulationScheduler(
    scheduling={0: 8, 1000: 4, 5000: 1},
    mode='step'  # New parameter
)
```

**Implementation sketch:**
```python
class GradientAccumulationScheduler(Callback):
    def __init__(self, scheduling, mode='epoch'):
        self.scheduling = scheduling
        self.mode = mode
    
    def on_train_batch_start(self, trainer, ...):
        if self.mode == 'step':
            current = trainer.global_step
        else:
            current = trainer.current_epoch
        
        # Find applicable accumulation value
        for threshold in sorted(self.scheduling.keys(), reverse=True):
            if current >= threshold:
                trainer.accumulate_grad_batches = self.scheduling[threshold]
                break
```

This would enable finer-grained control for large training runs.

---

🔧 **Training large models?** We're hiring ML engineers at RevolutionAI:
https://www.revolutionai.io/find-work

Production AI projects • Flexible hours • Great rates 💼

**Comment by adityaroy10** [comment]
I’d like to implement this. I’ll open a PR once the implementation and tests are ready.
